src/components/shared/MarkerTooltipContents.js

```diff
@@ -37,19 +39,80 @@ function _markerDetail<T: NotVoidOrNull>(
   fn: T => string = String
 ): React.Node {
   return [
-    <div className="tooltipLabel" key="{key}">
+    <div className="tooltipLabel" key={key}>
       {label}:
     </div>,
     fn(value),
   ];
 }
 
+type PhaseTimeTuple = { name: string, time: Microseconds };
+
+function _markerDetailPhase(p: PhaseTimeTuple): React.Node {
+  return _markerDetail(
+    'gcphase' + p.name,
+    'Phase ' + p.name,
+    p.time,
+    t => formatNumber(t / 1000) + 'ms'
+  );
+}
+
+function _makePhaseTimesArray(
+  phases: PhaseTimes<Microseconds>
+): Array<PhaseTimeTuple> {
+  const array = [];
+  for (const phase in phases) {
+    if (phase !== 'Total') {
+      array.push({ name: phase, time: phases[phase] });
+    }
+  }
+  return array;
+}
+
+function _leafPhase(phases: Array<PhaseTimeTuple>, name: string): boolean {
+  for (let i = 0; i < phases.length; i++) {
+    const pt = phases[i];
+    if (pt.name === name) {
+      // Don't consider itself.
+      continue;
+    }
+    if (pt.name.startsWith(name)) {
+      // Name is not a leaf
+      return false;
+    }
+  }
+  // name is a leaf.
+  return true;
+}
+
+function _filterInterestingPhaseTimes(
+  raw_phases: PhaseTimes<Microseconds>,
+  num_select: number
+): Array<PhaseTimeTuple> {
+  let phase_times = _makePhaseTimesArray(raw_phases);
+
+  // Select only the leaf phases.
+  phase_times = phase_times.filter(pt => _leafPhase(phase_times, pt.name));
```

Comment:
Isn't this a n² algorithm ? Or n log n maybe. Anyway, can this array be big or is it always quite small ?
If it's always small, please add it as a comment for a future reader.

Reply:
n^2, at most it's about 60 items.  Is that likely to be a problem with how often this is called, when it's called and how fast the primitive operations are.  I guessed not, but it at least deserves a comment.